server/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,132 @@
+package resourcemanager
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/mitchellh/mapstructure"
+)
+
+type Validator interface {
+	Validate() error
+}
+
+type Resource[T Validator] struct {
+	Type string `mapstructure:"type"`
+	Spec T      `mapstructure:"spec"`
+}
+
+type ResourceHandler[T Validator] interface {
+	Create(T) (T, error)
+}
+
+type manager[T Validator] struct {
+	resourceType string
+	handler      ResourceHandler[T]
+}
+
+func New[T Validator](resourceType string, handler ResourceHandler[T]) *manager[T] {
+	return &manager[T]{
+		resourceType: resourceType,
+		handler:      handler,
+	}
+}
+
+func (m *manager[T]) RegisterRoutes(r *mux.Router) *mux.Router {
+	// prefix is /{resourceType | lowercase}/
+	subrouter := r.PathPrefix("/" + strings.ToLower(m.resourceType)).Subrouter()
+
+	subrouter.HandleFunc("/", m.create).Methods(http.MethodPost)
+
+	return subrouter
+}
+
+func (m *manager[T]) create(w http.ResponseWriter, r *http.Request) {
+	encoder, err := encoderFromRequest(r)
+	if err != nil {
+		writeResponse(w, http.StatusBadRequest, fmt.Sprintf("cannot process request: %s", err.Error()))
+		return
+	}
+
+	values, err := readValues(r, encoder)
+	if err != nil {
+		writeResponse(w, http.StatusBadRequest, fmt.Sprintf("cannot parse body: %s", err.Error()))
+		return
+	}
+
+	targetResource := Resource[T]{}
+	err = mapstructure.Decode(values, &targetResource)
+	if err != nil {
+		writeResponse(w, http.StatusBadRequest, fmt.Sprintf("cannot unmarshal body values: %s", err.Error()))
+		return
+	}
+
+	// TODO: if resoruceType != values.resourceType return error
+
+	created, err := m.handler.Create(targetResource.Spec)
+	if err != nil {
+		writeResponse(w, http.StatusInternalServerError, fmt.Sprintf("error creating resource %s: %s", m.resourceType, err.Error()))
+		return
+	}
+
+	newResource := Resource[T]{
+		Type: m.resourceType,
+		Spec: created,
+	}
+
+	bytes, err := encodeValues(newResource, encoder)
+	if err != nil {
+		writeResponse(w, http.StatusInternalServerError, fmt.Sprintf("cannot marshal entity: %s", err.Error()))
+		return
+	}
+
+	writeResponse(w, http.StatusCreated, string(bytes))
+
+}
+
+func writeResponse(w http.ResponseWriter, code int, msg string) {
+	w.WriteHeader(code)
+	w.Write([]byte(msg))
```

Comment:
You probably need to write the Content-Type header here as well

Reply:
not strictly neccesary, but love the consistency this change introduces. great catch!